### PR TITLE
Start each request on fresh ExecutionContext

### LINKED ISSUE #13991: Kestrel is causing AsyncLocal to bleed context data across requests on same HTTP/1.1 connection
I have raised this issue with the coreclr repo but it seems to be caused by Kestrel as I cannot reproduce it with IISExpress or TestHost. 

https://github.com/dotnet/coreclr/issues/26713

Under the scenario described in the issue above and recreated in the provided repo, data stored within an AsyncLocal instance within a middleware is bleeding across requests.  

Is this a bug?

### PR DESCRIPTION
**Pull request by benaadams**
Check with Hide Whitespace on; as indents changed.

Fixes #13991

Per request allocations for Plaintext sample remains the same

Allocations for 17k requests across 256 connections

image

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs [hunk 1]
```diff
@@ -608,103 +615,124 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
...
+                // Start each request on a fresh ExecutionContext
+                ExecutionContext.Run(cleanExecutionContext, contextCallback, requestData);
+                // Await the request Task set in the callback; then clear it
+                await requestData.CurrentRequest;
+                requestData.CurrentRequest = default;
 
-                    // Run the application code for this request
-                    await application.ProcessRequestAsync(context);
+                // Clear the TContext for the request
+                application.DisposeContext(requestData.Context, _applicationException);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs]: Given that it's pluggable, should CreateContext and DisposeContext be called with the `cleanExecutionContext`?

Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs]: I suppose

Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs]: Done

**Comment by benaadams** [comment]
Per request allocations for Plaintext sample remains similar
image

**Comment by benaadams** [comment]
Cleaned up the encapsulation

**Comment by benaadams** [comment]
Since we are calling an extra async method anyway; realised it could be greatly simplified.

(ignore whitespace)

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs [hunk 2]
```diff
@@ -608,103 +610,111 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
...
+                if (HasStartedConsumingRequestBody)
                 {
-                    ReportApplicationError(ex);
+                    await messageBody.StopAsync();
                 }
+            }
+        }
 
-                KestrelEventSource.Log.RequestStop(this);
+        private async ValueTask ProcessRequest<TContext>(IHttpApplication<TContext> application)
```
Comment (benaadams) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs]: `Task` is a more performant return here than `ValueTask` and will allocate the same; i.e none in the sync completion and the async box in the async wait scenario (subject to https://github.com/dotnet/coreclr/pull/26310 when `ValueTask` becomes better as it doesn't allocate in async)

**Comment by halter73** [comment]
So much nicer! We probably still want to take the tests from #14027 though.

**Review by halter73 — APPROVED** [review]
One subtle difference between this and #14027 is that the Response.OnCompleted callback will run with the same ExecutionContext as the middleware if the middleware isn't "async". With #14027, the Response.OnCompleted gets a fresh ExecutionContext no matter what. I think this probably is fine though as long as we add a test and don't change behavior.

**Comment by benaadams** [comment]
> One subtle difference between this and #14027 is that the Response.OnCompleted callback will run with the same ExecutionContext as the middleware if the middleware isn't "async". 

Might even be desirable as it allows you to set a asynclocal (or "request" local) reference type in `OnStarted`; have it flow through the request and then be able to check its properties in `OnCompleted`

**Comment by benaadams** [comment]
Hmm... make that in the first middleware :)

**Comment by benaadams** [comment]
Added tests

> I think this probably is fine though as long as we add a test and don't change behaviour.

Also added compat test and commented it as such https://github.com/aspnet/AspNetCore/pull/14146/files#diff-d47b9baa7ec5338570136a89e5e8a4deR441-R444

```csharp
public async Task ExecutionContextMutationsOfValueTypeFlowThroughNonAsyncEvents()
{
    // Note: This test is for future compat behviour checking
    // to ensure non-async OnStarting/OnCompleted flow AsyncLocal changes out
```

**Comment by davidfowl** [comment]
We need to run performance tests

**Comment by benaadams** [comment]
> We need to run performance tests

Don't think it will be great :(

However... might as well make the OnStarting/OnComplete async as they won't regress performance; since normally aren't called?

**Comment by benaadams** [comment]
Changed it to flow normally for non-async events also; since they won't run if you haven't defined any so it shouldn't effect performance if you choose not to use them.

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs [hunk 3]
```diff
@@ -733,108 +743,55 @@ public void OnCompleted(Func<object, Task> callback, object state)
...
 
         protected Task FireOnCompleted()
         {
             var onCompleted = _onCompleted;
-
-            if (onCompleted == null || onCompleted.Count == 0)
+            if (onCompleted?.Count > 0)
             {
-                return Task.CompletedTask;
+                return ProcessEvents(this, onCompleted);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs]: If one of the other onCompleted callbacks throws, do we still continue looping over the other callbacks and executing? Unlike the onStarting callbacks, it's important that every onCompleted callback is invoked. I thought we had a test for this.

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs]: Nvm. I just saw this was calling a local function.

**Comment by anurse** [comment]
@halter73 are you good to merge this in for 5.0?

**Comment by davidfowl** [comment]
> @halter73 are you good to merge this in for 5.0?

Not without performance tests.

**Comment by benaadams** [comment]
Is that auto perf test ci thing working? https://github.com/aspnet/AspNetCore/pull/16999 (seems to be having versioning issues)

**Comment by davidfowl** [comment]
Maybe we can use the in memory unit tests.

**Comment by benaadams** [comment]
Lemme rebase

**Comment by benaadams** [comment]
Changed to ValueTask since https://github.com/dotnet/coreclr/pull/26310 was merged; but would be good to also test the perf with the env variable `DOTNET_SYSTEM_THREADING_POOLASYNCVALUETASKS` set to 1 to see have that changes things; probably on something with some async (database calls?) and then feedback to https://github.com/dotnet/coreclr/issues/27403

**Comment by anurse** [comment]
> Is that auto perf test ci thing working?

@halter73 says soon, so we'll try running it on this when it's working.

**Comment by anurse** [comment]
@aspnet-hello benchmark

**Comment by anurse** [comment]
@halter73 :(

**Comment by halter73** [comment]
@aspnet-hello benchmark

**Comment by benaadams** [comment]
329,492 -> 354,384?

Even when I make it slower I make it faster? 😅😎

**Comment by davidfowl** [comment]
Whattttt

**Comment by davidfowl** [comment]
@aspnet-hello benchmark

**Comment by benaadams** [comment]
@aspnet-hello benchmark

**Comment by benaadams** [comment]
😢

**Comment by benaadams** [comment]
Is there a way to test with the environment variable `DOTNET_SYSTEM_THREADING_POOLASYNCVALUETASKS` set to `1`?

https://github.com/dotnet/coreclr/issues/27403

**Comment by anurse** [comment]
Seems like this is within the noise. Are we ok taking it and monitoring the rolling benchmarks? @halter73 @davidfowl .

**Comment by benaadams** [comment]
@aspnet-hello benchmark

**Comment by anurse** [comment]
@davidfowl are you good with the perf on this? I'd like to close it out and preferably by merging it ;P

**Comment by anurse** [comment]
Ping @davidfowl @halter73

**Comment by anurse** [comment]
Hearing nothing from @davidfowl, I'm going to go ahead and tell @halter73 he can merge away ;).

**Comment by davidfowl** [comment]
Merge it we can measure things now

**Comment by benaadams** [comment]
Rebased

**Comment by benaadams** [comment]
@aspnet-hello benchmark

**Comment by benaadams** [comment]
@aspnet-hello benchmark

**Comment by BIOsborne** [comment]
I have observed an issue (on AspNetCore version 3.1.13) with AsyncLocal context being leaked between requests. (Never in parallel, the Context from one request which finishes appears to be present at the start of a subsequent one).
  
I have not attempted (yet) to reproduce this locally, but have observed it reliably in our Production systems.
Interestingly - we are running 50% production traffic through Kestrel (running on Linux) and 50% running through Azure Windows hosted with IIS (as part of a live migration process). The issue appears only with Kestrel on Linux. Not sure if that is relevant.
  
I have seen this issue is closed now: https://github.com/dotnet/aspnetcore/issues/13991
I assume we would not expect this issue to be live still?
If this is of interest, I can create a new issue, but won't do this if there are live issues already open around this that I have not been able to find.

A little more context - the AsyncLocal state tracked in our case gets initialised in Log sink which is NOT in an Async method. (The only reason I mention that is that some comments in the linked issue refer to the method setting the AsyncLocal state should be async.
(https://github.com/dotnet/aspnetcore/issues/13991#issuecomment-531633911)

Any input gratefully received - hope this isn't wasting your time!
Thanks.

**Comment by alefranz** [comment]
@BIOsborne this fix is in .NET 5 and has not been backported to 3.1.
As far as I am aware, the only way to work around the problem in 3.1 is what David suggested in the comment you pointed out.

**Comment by BIOsborne** [comment]
> @BIOsborne this fix is in .NET 5 and has not been backported to 3.1.
> As far as I am aware, the only way to work around the problem in 3.1 is what David suggested in the comment you pointed out.

Thanks @alefranz for quick reply - good to know.
We have been able to work around this in 3.1.* for now no problem.

**Pull request merged**